### ISSUE TITLE
Handle response failed: Broken pipe

### ISSUE THREAD

**Issue by jpecar** [ethereum-mining/ethminer#171]
Using 0.11. Mining works fine, but in the past week or so I started seeing these more and more:

```
  ?  12:23:07|stratum  Received new job #0x6ee1fd
  ?  12:23:07|openclminer0  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:23:07|openclminer1  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:23:07|openclminer2  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:23:07|openclminer3  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:24:03|openclminer1  Solution found; Submitting to eu1.ethermine.org ...
  ?  12:24:03|openclminer1    Nonce: 0x0ef54ac32a91d04c
  ?  12:24:24|openclminer5  Solution found; Submitting to eu1.ethermine.org ...
  ?  12:24:24|openclminer5    Nonce: 0x2feb7b930ea30c49
  ?  12:24:24|openclminer5  Submitting stale solution.
  ?  12:24:34|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ?  12:24:34|openclminer3    Nonce: 0x712378386dd2d76b
  ?  12:25:38|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ?  12:25:38|openclminer3    Nonce: 0x71237838d4f3f0b1
  ?  12:26:01|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ?  12:26:01|openclminer3    Nonce: 0x71237838fa895fe8
  ?  12:26:43|openclminer4  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:26:43|openclminer5  set work; seed: #68fc811a, target:  #0000000112e0
  ?  12:26:43|stratum  Handle response failed: Broken pipe
  ?  12:26:43|stratum  Handle response failed: Broken pipe
  ?  12:26:43|stratum  Handle response failed: Broken pipe
  ?  12:26:43|stratum  Handle response failed: Broken pipe
  ?  12:26:43|stratum  Handle response failed: Broken pipe
  ?  12:26:43|stratum  :-( Not accepted.
  ?  12:26:43|stratum  :-( Not accepted.
```

"Broken pipe" pile up for some hours and at the end miner crashes with "Unhandled signal in divisionErrorHandler()".

I assume this is somehow related to pools and their ability to process requests. I'm seeing this with ethermine.org and ethpool.org.

**Comment by derubm** [comment]
ethminer --help shows : 
 
-SC, --stratum-client <n>  Stratum client version. Defaults to 1 (async client). Use 2 to use the new synchronous client.
    -SP, --stratum-protocol <n> Choose which stratum protocol to use:
        0: official stratum spec: ethpool, ethermine, coinotron, mph, nanopool (default)
        1: eth-proxy compatible: dwarfpool, f2pool, nanopool
        2: EthereumStratum/1.0.0: nicehash
did you try the corresponding  -SP parameter for your pool ?

**Comment by jpecar** [comment]
So ... above is with -SP 1. This happens with -SP 0 -SC 2 after about 2h:
```
  ℹ  18:01:14|stratum  Received new job #c4f6d4b4
  ℹ  18:01:14|openclminer0  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:01:15|openclminer1  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:01:15|openclminer2  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:01:15|openclminer3  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:01:40|openclminer5  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:01:40|openclminer5    Nonce: 0x09e487ce1df51a39
  ✘  18:01:40|openclminer5  Submitting stale solution.
  ℹ  18:02:06|openclminer2  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:02:06|openclminer2    Nonce: 0xceea02f1aac38cb2
  ℹ  18:03:00|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:03:00|openclminer3    Nonce: 0x5755fb3a099697cd
  ℹ  18:03:01|openclminer1  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:03:01|openclminer1    Nonce: 0x7519cef77fda565e
  ℹ  18:03:15|openclminer0  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:03:15|openclminer0    Nonce: 0x0779e967c6d8aa9d
  ℹ  18:04:08|openclminer5  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:04:08|openclminer5    Nonce: 0x09e487cf1dc9510f
  ✘  18:04:08|openclminer5  Submitting stale solution.
  ℹ  18:05:34|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:05:34|openclminer3    Nonce: 0x5755fb3b0d0054fc
  ℹ  18:05:59|openclminer0  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:05:59|openclminer0    Nonce: 0x0779e968e52dd158
  ℹ  18:06:00|openclminer0  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:06:00|openclminer0    Nonce: 0x0779e968e5e65b2c
  ✘  18:06:00|openclminer0  Exception thrown in Worker thread: write: Broken pipe
  ℹ  18:07:19|openclminer3  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:07:19|openclminer3    Nonce: 0x5755fb3bba68db00
  ✘  18:07:19|openclminer3  Exception thrown in Worker thread: write: Broken pipe
  ℹ  18:08:25|openclminer4  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:08:25|openclminer5  set work; seed: #68fc811a, target:  #0000000112e0
  ℹ  18:08:25|stratum  Received new job #1300c75e
  ℹ  18:09:44|openclminer1  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:09:44|openclminer1    Nonce: 0x7519cef9eb5284c8
  ✘  18:09:44|openclminer1  Submitting stale solution.
  ✘  18:09:44|openclminer1  Exception thrown in Worker thread: write: Broken pipe
  ℹ  18:11:55|openclminer2  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:11:55|openclminer2    Nonce: 0xceea02f5a3654899
  ✘  18:11:55|openclminer2  Submitting stale solution.
  ✘  18:11:55|openclminer2  Exception thrown in Worker thread: write: Broken pipe
  ℹ  18:20:47|openclminer5  Solution found; Submitting to eu1.ethermine.org ...
  ℹ  18:20:47|openclminer5    Nonce: 0xc3cff4e45b362aaa
  ✘  18:20:47|openclminer5  Submitting stale solution.
  ✘  18:20:47|openclminer5  Exception thrown in Worker thread: write: Broken pipe
```
And it is now stuck here for almost 1h.

**Comment by jpecar** [comment]
Same happens with -SP 2 and nicehash. After about 2h I get "Unhandled signal in divisionErrorHandler()"

**Comment by jpecar** [comment]
-SP 2 with -SC 2 to nicehash also deadlocks after a while ...

**Comment by jpecar** [comment]
Using lsof and netstat I see that ethminer process has a connection established to pool stratum server. At the moment I start seeing issues in its output, I see this connection in close_wait and then it goes away completely. But ethminer is not aware of that and is still trying to write to the same socket, which netstat shows as "can't identify protocol". Also ethminer doesn't try to reconnect or to switch to backup server, defined by -FS.

**Comment by derubm** [comment]
If you guys like i can hand out a temporary Solution based on the error message, so your miner would restart - permanent fix should be done in miner or pool end of course.

**Comment by DeadManWalkingTO** [comment]
I believe now the issue have been solved #757 (added --exit parameter to exit whenever an error occurred).

Check #735 (Simple Script WatchDog).

**Comment by smurfy** [comment]
Well, thats an stratum error, not an gpu error, so its not handled with the exit param.
But stratum and pools in general got refactored quite recently so maybe this is fixed anyways?
